File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/topk_softmax_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/topk_softmax_emit.py -- Python reference emitter for the
# topk-softmax parity harness. Selects one of the sampled configs by argv[1]
# (the config index), builds the TopkSoftmaxSpec, builds the kernel via
# build_topk_softmax and prints lower_kernel_to_llvm(arch='gfx950') to stdout
# so it can be byte-compared with the C emitter topk_softmax_emit.c.
from rocke.instances.common.topk_softmax import (
    TopkSoftmaxSpec,
    build_topk_softmax,
)
from _emit_common import run_emit


def _spec(idx: int) -> TopkSoftmaxSpec:
    if idx == 0:
        return TopkSoftmaxSpec(
            n_per_row=32, k=1, dtype="f32", out_dtype="f32", block_size=32
        )
    if idx == 1:
        return TopkSoftmaxSpec(
            n_per_row=64, k=4, dtype="f16", out_dtype="f32", block_size=64
        )
    if idx == 2:
        return TopkSoftmaxSpec(
            n_per_row=128, k=8, dtype="bf16", out_dtype="bf16", block_size=64
        )
    if idx == 3:
        return TopkSoftmaxSpec(
            n_per_row=4096, k=16, dtype="f32", out_dtype="f32", block_size=128
        )
    if idx == 4:
        return TopkSoftmaxSpec(
            n_per_row=16384, k=32, dtype="f32", out_dtype="f32", block_size=256
        )
    if idx == 5:
        return TopkSoftmaxSpec(
            n_per_row=768, k=2, dtype="f32", out_dtype="f32", block_size=64
        )
    raise SystemExit(f"unknown config index {idx}")


def main() -> int:
    return run_emit(
        _spec,
        build_topk_softmax,
        usage="usage: topk_softmax_emit.py <config_index 0..5> [mode]\n",
    )


if __name__ == "__main__":
    raise SystemExit(main())

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/topk_softmax_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/topk_softmax_emit.c -- C-side emitter for the topk-softmax
 * instance parity harness. Selects one of the sampled configs by argv[1] (the
 * config index), builds rocke_topk_softmax_spec_t identically to the Python
 * emitter topk_softmax_emit.py, validates via rocke_topk_softmax_is_valid_spec,
 * builds into a fresh IRBuilder via rocke_build_topk_softmax_new (the C build
 * entry), lowers via rocke_lower_kernel_to_llvm (arch gfx950, flavor AUTO) and
 * prints the .ll to stdout so the two outputs can be byte-compared.
 *
 * Optional argv[2] = mode: "ll" (default), "ir", "verify".
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_topk_softmax.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, rocke_topk_softmax_spec_t* spec)
{
    *spec = rocke_topk_softmax_spec_default();

    switch(idx)
    {
    case 0:
        spec->n_per_row = 32;
        spec->k = 1;
        spec->dtype = "f32";
        spec->out_dtype = "f32";
        spec->block_size = 32;
        break;
    case 1:
        spec->n_per_row = 64;
        spec->k = 4;
        spec->dtype = "f16";
        spec->out_dtype = "f32";
        spec->block_size = 64;
        break;
    case 2:
        spec->n_per_row = 128;
        spec->k = 8;
        spec->dtype = "bf16";
        spec->out_dtype = "bf16";
        spec->block_size = 64;
        break;
    case 3:
        spec->n_per_row = 4096;
        spec->k = 16;
        spec->dtype = "f32";
        spec->out_dtype = "f32";
        spec->block_size = 128;
        break;
    case 4:
        spec->n_per_row = 16384;
        spec->k = 32;
        spec->dtype = "f32";
        spec->out_dtype = "f32";
        spec->block_size = 256;
        break;
    case 5:
        spec->n_per_row = 768;
        spec->k = 2;
        spec->dtype = "f32";
        spec->out_dtype = "f32";
        spec->block_size = 64;
        break;
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index 0..5> [mode]\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    if(strcmp(mode, "ll") != 0 && strcmp(mode, "ir") != 0 && strcmp(mode, "verify") != 0)
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        return 2;
    }

    rocke_topk_softmax_spec_t spec;
    if(make_spec(idx, &spec) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    const char* arch = "gfx950";

    /* Validate the spec (mirrors is_valid_spec). */
    char reason[ROCKE_ERR_MSG_CAP];
    reason[0] = 0;
    if(!rocke_topk_softmax_is_valid_spec(&spec, arch, reason, sizeof reason))
    {
        fprintf(stderr, "invalid spec: %s\n", reason);
        return 1;
    }

    /* Init IRBuilder with spec.kernel_name() and build into it. */
    rocke_ir_builder_t b;
    rocke_kernel_def_t* kernel = rocke_build_topk_softmax_new(&b, &spec, arch);
    if(!kernel)
    {
        fprintf(stderr, "build failed: %s\n", b.err);
        rocke_ir_builder_free(&b);
        return 1;
    }

    if(strcmp(mode, "ir") == 0)
    {
        char* t = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &t);
        if(st != ROCKE_OK || !t)
        {
            fprintf(stderr, "ir_serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
    }
    else if(strcmp(mode, "verify") == 0)
    {
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }
    else
    {
        /* mode == "ll" */
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, arch, &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    rocke_ir_builder_free(&b);
    return 0;
}
